### DISCUSSION TITLE
Some way to pause or end a recording from within the recorded process?

### DISCUSSION THREAD

**Discussion by moyix** [asciinema/discussions#241] [Ideas]
I'm currently using asciinema to record all of my terminal sessions. However, I would like to be able to programmatically stop or pause a recording; for example, when ssh-ing to another system (so that only my local activity gets recorded). Killing the parent asciinema process is either ignored (if I use HUP/TERM/etc) or causes the shell itself to be recorded (if I use SIGKILL).

For pausing I have figured out a somewhat horrifying workaround. Assuming you have permissions (it requires root on macOS), you can send characters directly to another terminal's stdin by using the `TIOCSTI` ioctl on the terminal device (e.g., `/dev/ttys002`). The program `writevt` ([source available here](https://github.com/grawity/code/blob/main/thirdparty/writevt.c); just use `gcc writevt.c -o writevt` to compile) can be used to do this from the command line.

So here's a function I added to my bash startup script. Now inside a recording I can call `asciinema_toggle_pause` and it will pause the recording for me. Handling more deeply nested subshells is ~~left as an exercise for the reader~~ also handled.

```bash
function asciinema_toggle_pause {
    # Change this if your config uses a different pause char. The default is
    # Ctrl-\; you can convert by taking ord(c)-0x40, where c is the character
    # after Ctrl-. You might also need to add a leader char if that's configured.
    PAUSE_CHAR=$(echo -ne '\x1c')

    # If it's not running, don't do anything
    if [ -z "${ASCIINEMA_REC+x}" ]; then
        return
    fi

    # Walk up the process tree until we find the asciinema process
    # Get the ppid of the current process
    current_pid=$$
    ppid=$(ps -o ppid= $current_pid)
    # Get the command name to check
    cmdname=$(ps -o command= $ppid)
    # If the command name contains "asciinema", then we're good
    while [[ $cmdname != *"asciinema"* ]]; do
        # If we've reached the top of the tree, then asciinema isn't running
        if [[ $ppid == "1" ]]; then
            echo "WARNING: asciinema is not running (but ASCIINEMA_REC is set)"
            return
        fi
        # Otherwise, keep going
        current_pid=$ppid
        ppid=$(ps -o ppid= $current_pid)
        cmdname=$(ps -o command= $ppid)
    done

    # Get its tty. xargs is used to trim whitespace
    ttyname=/dev/$(ps -o tty= $ppid | xargs)

    # Inject the pause char
    # echo "Pausing/unpausing asciinema on tty $ttyname (pid $ppid)"
    sudo writevt "$ttyname" "$PAUSE_CHAR"
}
```

I also added writevt to `sudoers` for my user so that it doesn't need a password. This is horribly insecure on a multiuser system, since it means I'd have the ability to inject commands into anyone else's terminal sessions. But this is just my laptop so it's fine:

`/private/etc/sudoers.d/001-writevt`:
```sudoers
moyix ALL=(ALL) NOPASSWD: /Users/moyix/bin/writevt
```

**Comment by ku1ik** [comment]
Nice! This is some serious unix hacking :clap: 

Would it make it simpler if asciinema had some sort of RPC mechanism to toggle pause? E.g. by invoking something like `asciinema rpc pause` from within recording session, which would automatically find related asciinema process and make it pause? Similar to how you can send keys to tmux with `tmux send-keys ...`.

**Reply by moyix** [reply]
Yes, I think that would be perfect. I actually hacked together a little control interface that uses a socket file last night; it works slightly differently though:

* When starting the recording, it sets a new environment variable `ASCIINEMA_CTL`, which has a path to the socket.
* The fd monitoring (`copy` in `pty_.py`) also monitors for connections on the control socket and activity on those connections, and calls a handler for the control commands.
* There's a new command `asciinema control` that connects to the socket in `ASCIINEMA_CTL` and sends a command and prints the response (if any).
* The set of commands available right now is `pause`, `resume` `mark`, and `status`, but of course others could be added.

You can take a look at the code here; happy to create a PR if there's interest! https://github.com/asciinema/asciinema/compare/develop...moyix:asciinema:develop

Demo of using it: https://asciinema.org/a/J2xzBmySyful3n3tbudk6tcKI

And the resulting recorded cast: https://asciinema.org/a/wpVnBP7EtW522nr47TVhxZARX

**Reply by ku1ik** [reply]
Great, this seems useful. Regarding incorporating this into asciinema, I'm currently rewriting it in Rust (https://github.com/asciinema/asciinema/tree/next/src) :sweat_smile: This is going to be 3.0 release. The current Python codebase is not going to be extended with any new features, only with major bugfixes for existing functionality. We can revisit this idea when 3.0 is out and stable.

**Comment by ku1ik** [comment]
Another alternative to the control socket above could be control via signals. asciinema could set `ASCIINEMA_PID` env var when starting  recorded process, e.g. shell, and then one could do pause/resume via sth like `kill -USR1 $ASCIINEMA_PID` or similar. The set of available signals is not very wide but this solution is nice in its simplicity IMHO.
